README
NAME
    Message::Passing - a simple way of doing messaging.

SYNOPSIS
        message-pass --input STDIN --output STDOUT
        {"foo": "bar"}
        {"foo":"bar"}

DESCRIPTION
    A library for building high performance, loosely coupled and
    reliable/resilient applications, structured as small services which
    communicate over the network by passing messages.

  BASIC PREMISE
    You have data for discrete events, represented by a hash (and serialized
    as JSON).

    This could be a text log line, an audit record of an API event, a metric
    emitted from your application that you wish to aggregate and process -
    anything that can be a simple hash really..

    You want to be able to shove these events over the network easily, and
    aggregate them / filter and rewrite them / split them into worker
    queues.

    This module is designed as a simple framework for writing components
    that let you do all of these things, in a simple and easily extensible
    manor.

    For a practical example, You generate events from a source (e.g. ZeroMQ
    output of logs and performance metrics from your Catalyst FCGI or
    Starman workers) and run one script that will give you a central
    application log file, or push the logs into Elasticsearch.

    There are a growing set of components you can plug together to make your
    solution.

    Getting started is really easy - you can just use the "message-passing"
    command installed by the distribution. If you have a common config that
    you want to repeat, or you want to write your own server which does
    something more flexible than the normal script allows, then see
    Message::Passing::DSL.

    To dive straight in, see the documentation for the command line utility
    message-pass, and see the examples in
    Message::Passing::Manual::Cookbook.

    For more about how the system works, see
    Message::Passing::Manual::Concepts.

COMPONENTS
    Below is a non-exhaustive list of components available.

  INPUTS
    Inputs receive data from a source (usually a network protocol).

    They are responsible for decoding the data into a hash before passing it
    onto the next stage.

    Inputs include:

    Message::Passing::Input::STDIN
    Message::Passing::Input::ZeroMQ
    Message::Passing::Input::STOMP
    Message::Passing::Input::AMQP
    Message::Passing::Input::Syslog
    Message::Passing::Input::Redis
    Message::Passing::Input::Test

    You can easily write your own input, just use AnyEvent, and consume
    Message::Passing::Role::Input.

  FILTER
    Filters can transform a message in any way.

    Examples include:

    Message::Passing::Filter::Null - Returns the input unchanged.
    Message::Passing::Filter::All - Stops any messages it receives from
    being passed to the output. I.e. literally filters all input out.
    Message::Passing::Filter::T - Splits the incoming message to multiple
    outputs.

    You can easily write your own filter, just consume
    Message::Passing::Role::Filter.

    Note that filters can be chained, and a filter can return undef to stop
    a message being passed to the output.

  OUTPUTS
    Outputs send data to somewhere, i.e. they consume messages.

    Message::Passing::Output::STDOUT
    Message::Passing::Output::AMQP
    Message::Passing::Output::STOMP
    Message::Passing::Output::ZeroMQ
    Message::Passing::Output::WebHooks
    Message::Passing::Output::Search::Elasticsearch
    Message::Passing::Output::Redis
    Message::Passing::Output::Test

SEE ALSO
    Message::Passing::Manual - The manual (contributions cherished).
    <http://www.slideshare.net/bobtfish/messaging-interoperability-and-log-a
    ggregation-a-new-framework> - Slide deck!
    Log::Message::Structured - For creating your log messages.
    Log::Dispatch::Message::Passing - use Message::Passing outputs from
    Log::Dispatch.

THIS MODULE
    This is a simple MooX::Options script, with one input, one filter and
    one output. To build your own similar scripts, see:

    Message::Passing::DSL - To declare your message chains.
    Message::Passing::Role::CLIComponent - To provide "foo" and
    "foo_options" attribute pairs.
    Message::Passing::Role::Script - To provide daemonization features.

  METHODS
   build_chain
    Builds and returns the configured chain of input => filter => output.

   start
    Class method to call the run_message_server function with the results of
    having constructed an instance of this class, parsed command line
    options and constructed a chain.

    This is the entry point for the script.

AUTHOR
    Tomas (t0m) Doran <bobtfish@bobtfish.net>

SUPPORT
  Bugs
    Please log bugs at rt.cpan.org. Each distribution has a bug tracker link
    in its metacpan.org page.

  Discussion
    #message-passing on irc.perl.org.

  Source code
    Source code for all modules is available at <http://github.com/suretec>
    and forks/patches are very welcome.

SPONSORSHIP
    This module exists due to the wonderful people at Suretec Systems Ltd.
    <http://www.suretecsystems.com/> who sponsored its development for its
    VoIP division called SureVoIP <http://www.surevoip.co.uk/> for use with
    the SureVoIP API -
    <http://www.surevoip.co.uk/support/wiki/api_documentation>

COPYRIGHT
    Copyright Suretec Systems Ltd. 2012.

    Logstash (upon which many ideas for this project are based, but from
    which we do not reuse any code) is copyright 2010 Jorden Sissel.

LICENSE
    GNU Library General Public License, Version 2.1



Changes
  - Add missing List::MoreUtils dep
  - Fix missing JSON dependency
    by using JSON::MaybeXS also for tests
  - Document Output::Test clear_messages method

0.116 2015-02-23T18:00:00+01:00
  - Fix link in 'Writing Filters' section of
    Message::Passing::Manual::Components
  - Fix daemonize_if_needed spelling
  - Add documentation for all command line options
  - Add Linux::IO_Prio to recommended modules for io_priority option
  - Recommend newer Message::Passing::Output::Search::Elasticsearch in docs

0.115 2014-08-23T09:55:00+03:00 YAPC::EU 2014 release
  - Fix internal and therefore undocumented type parameter swallows
    Output::Search::Elasticsearches type parameter
  - Use JSON::MaybeXS instead of JSON in Filter::De/Encoder::JSON

0.114 2013-09-17T10:30:00+02:00
  - Fix test failures under Perl 5.18
  - Improve documentation

0.113 2013-08-02T09:25:00+02:00
  - Add Message::Passing::Filter::Mangle for use with Message::Passing::DSL
    when you don't want to write a named filter.

0.112 2013-04-25T23:45:39+0100
  - Change to LGPL_2_1.

0.111 2013-03-03T15:54:12+0000
  - Fix STDIN input to exit on EOF.

0.110 2012-10-08T16:51:49+0100
  - Fix v-string to work with perl 5.8. RT#80034
  - Fix issues stopping multiple FileTail inputs being used in the
    same process.

0.109 2012-09-27T16:11:50+0100
  - Really fix test fail :(

0.108 2012-09-27T14:59:43+0100
  - Fix test fail on some machines.

0.107 2012-09-27T10:03:30+0100
  - Change Input::FileTail to use Sys::Hostname::Long

  - Fix bug causing crash with no helpful error message if PID file
    location is not writeable.

0.106 2012-09-12T23:59:13+0100
  - Remove debug warning accidentally left in the UDP input.

0.105 2012-09-12T23:55:36+0100
  - Depend on latest MooX::Options to stop install fails.

0.104 2012-09-06T18:36:27+0200
  - Fix to work with MooX::Options 3.7

0.103 2012-09-05T09:07:00+0100
  - Message::Passing::Input::FileTail now defaults to outputting a hash
    of data which contains additional metadata to the message line,
    such as the hostname, timestamp, filename. A 'raw => 1' option has
    been added which can be used to restore the original behaviour.

  - Add an experimental (and undocumented)
    Message::Passing::Filter::ToLogstash

  - Fix issue in JSON decoder when processing exceptions.

0.102 2012-08-26T21:12:52+0100
  - Add error chain support to the JSON encoder and decoder so that an
    error is logged if JSON encoding or decoding fails.

  - Make default error chain do JSON encoding on errors so that they're
    readable.

  - Make CLI error chain have options for the error encoder.

0.101 2012-08-23T21:23:52+0100
  - Fix daemonization features provided by Message::Passing::Role::Script
    to work on the command line again by exlicitly using MooX::Options

  - Fix Message::Passing::Input::FileTail to not die if the tail process
    is killed for any reason. It is instead re-spawned.. RT#78851

0.100 2012-08-19T08:12:32+0100
  - Port the dist to Moo, so that the core of Message::Passing
    uses no XS code. Components can still be written using full
    Moose, as Moo supports upgrading classes to full Moose classes.

    Note however that this implies the following changes to custom scripts:

    - Scripts should use MooX::Options instead of MooseX::Getopt,
      attributes will need to be updated to use the 'option' keyword
      rather than 'has'.

    - Message::Passing::Role::CLIComponent is now built on top of
      Package::Variant rather than MooseX::Role::Parameterized,
      and so scripts should now import it and use the function
      provided to generate roles.

  - Add UDP socket Input and Output, allowing transit of arbitrary
    UDP packets.

  - Add Travis continuous integration to the project.

0.010 2012-07-11T19:17:44+0100
  - Doc fix in Message::Passing::Manual::Cookbook
    https://github.com/suretec/Message-Passing/pull/2

  - Add Crypt::CBC encoder and decoder.

  - Add Gzip and Bzip2 encoders and decoders.

0.009 2012-06-13T15:53:35-0500
  - Add Message::Passing::Manual::Components.

  - Add SYNOPSIS to Message::Passing::Role::Filter.

  - Additional documentation in Message::Passing::Manual.

  - Refactor building connection managers so less duplicate code
    is needed in components.

  - Make STDIN input saner.

  - Add an 'error chain' concept, allowing components to have a
    way of reporting connection (or other) issues out of band
    with the message stream.
    This defaults to STDERR, however can be overridden per component,
    or for a whole chain.

  - Add error_log() function to the DSL for setting the error chain up
    for a whole chain.

  - Use the error chain in Message::Passing::Role::ConnectionManager
    to output errors when connections disconnect / timeout / reconnect.

  - Cleanups to Message::Passing::Input::FileTail to not leak processes
    and file handles if the instance of the Input class is destroyed.

0.008 2012-06-10T20:51:51+0100
  - Add standard roles to unify the names of connection attributes:
    Message::Passing::Role::HasHostnameAndPort
    Message::Passing::Role::HasUsernameAndPassword

  - Lots of additional documentation.

  - Add STDERR output.

  - Rename the log_chain function to message_chain and run_log_server
    function to run_message_server, in Message::Passing::DSL
    to follow the general rename, as we're not just about logs.

0.007 2012-06-10T11:07:45+0100
  - Documentation in the message-pass script

  - Add --configfile option to default script, allowing
    you to load config from a file, rather than supplying
    it on command line.

  - Make JSON encoder pass non refs straight through,
    so that if a previous filter generates a scalar,
    then this gets sent as-is.

  - Make JSON decoder pass refs straight through, to
    act as a no-op if the input has already decoded
    its data into a hash.

  - Remove spurious warnings from reconnect code.

0.006 2012-06-08T01:30:53+0100

lib/Message/Passing.pm
package Message::Passing;
use Moo;
use Config::Any;
use Message::Passing::Role::CLIComponent;
use Message::Passing::DSL;
use Carp qw/ confess /;
use MooX::Options flavour => [qw( pass_through )], protect_argv => 0;
use namespace::clean -except => [qw/ meta new_with_options parse_options _options_data _options_config/];
use 5.008004;

our $VERSION = '0.116';
$VERSION = eval $VERSION;

around 'parse_options' => sub {
    my $orig = shift;
    my $class = shift;
    my %args = $orig->($class, @_);

    if (my $conf = $args{configfile}) {
        my $cfg = $class->get_config_from_file($conf);
        foreach my $k (keys %$cfg) {
            if (!exists $args{$k}) {
                $args{$k} = $cfg->{$k};
            }
        }
    }

    return %args;
};


with
    CLIComponent( name => 'input' ),
    CLIComponent( name => 'output' ),
    CLIComponent( name => 'filter', default => 'Null' ),
    CLIComponent( name => 'decoder', default => 'JSON' ),
    CLIComponent( name => 'encoder', default => 'JSON' ),
    CLIComponent( name => 'error', default => 'STDERR' ),
    CLIComponent( name => 'error_encoder', default => 'Message::Passing::Filter::Encoder::JSON' ),
    'Message::Passing::Role::Script';

option configfile => (
    is => 'ro',
    format => 's',
);

sub get_config_from_file {
    my ($class, $filename) = @_;
    my ($fn, $cfg) = %{ Config::Any->load_files({
        files => [$filename],
        use_ext => 1,
    })->[0] };
    return $cfg;
}

sub build_chain {
    my $self = shift;
    message_chain {
        error_log(
            %{ $self->error_encoder_options },
            class => $self->error_encoder,
            output_to => output error => (
                %{ $self->error_options },
                class => $self->error,
            ),
        );
        output output => (
            %{ $self->output_options },
            class => $self->output,
        );
        encoder("encoder",
            %{ $self->encoder_options },
            class => $self->encoder,
            output_to => 'output',
        );
        filter filter => (
            %{ $self->filter_options },
            class => $self->filter,
            output_to => 'encoder',
        );
        decoder("decoder",
            %{ $self->decoder_options },
            class => $self->decoder,
            output_to => 'filter',
        );
        input input => (
            %{ $self->input_options },
            class => $self->input,
            output_to => 'decoder',
        );
    };
}

1;

=head1 NAME

Message::Passing - a simple way of doing messaging.

=head1 SYNOPSIS

    message-pass --input STDIN --output STDOUT
    {"foo": "bar"}
    {"foo":"bar"}

=head1 DESCRIPTION

A library for building high performance, loosely coupled and reliable/resilient applications,
structured as small services which communicate over the network by passing messages.

=head2 BASIC PREMISE

You have data for discrete events, represented by a hash (and
serialized as JSON).

This could be a text log line, an audit record of an API
event, a metric emitted from your application that you wish
to aggregate and process - anything that can be a simple hash really..

You want to be able to shove these events over the network easily,
and aggregate them / filter and rewrite them / split them into worker queues.

This module is designed as a simple framework for writing components
that let you do all of these things, in a simple and easily extensible
manor.

For a practical example, You generate events from a source (e.g.
L<ZeroMQ> output of logs and performance metrics from your L<Catalyst> FCGI
or L<Starman> workers) and run one script that will give you a central
application log file, or push the logs into Elasticsearch.

There are a growing set of components you can plug together
to make your solution.

Getting started is really easy - you can just use the C<message-pass>
command installed by the distribution. If you have a common config
that you want to repeat, or you want to write your own server
which does something more flexible than the normal script allows,
then see L<Message::Passing::DSL>.

To dive straight in, see the documentation for the command line utility
L<message-pass>, and see the examples in L<Message::Passing::Manual::Cookbook>.

For more about how the system works, see L<Message::Passing::Manual::Concepts>.

=head1 COMPONENTS

Below is a non-exhaustive list of components available.

=head2 INPUTS

Inputs receive data from a source (usually a network protocol).

They are responsible for decoding the data into a hash before passing
it onto the next stage.

Inputs include:

=over

=item L<Message::Passing::Input::STDIN>

=item L<Message::Passing::Input::ZeroMQ>

=item L<Message::Passing::Input::STOMP>

=item L<Message::Passing::Input::AMQP>

=item L<Message::Passing::Input::Syslog>

=item L<Message::Passing::Input::Redis>

=item L<Message::Passing::Input::Test>

=back

You can easily write your own input, just use L<AnyEvent>, and
consume L<Message::Passing::Role::Input>.

=head2 FILTER

Filters can transform a message in any way.

Examples include:

=over

=item L<Message::Passing::Filter::Null> - Returns the input unchanged.

=item L<Message::Passing::Filter::All> - Stops any messages it receives from being passed to the output. I.e. literally filters all input out.

=item L<Message::Passing::Filter::T> - Splits the incoming message to multiple outputs.

=back

You can easily write your own filter, just consume
L<Message::Passing::Role::Filter>.

Note that filters can be chained, and a filter can return undef to
stop a message being passed to the output.

=head2 OUTPUTS

Outputs send data to somewhere, i.e. they consume messages.

=over

=item L<Message::Passing::Output::STDOUT>

=item L<Message::Passing::Output::AMQP>

=item L<Message::Passing::Output::STOMP>

=item L<Message::Passing::Output::ZeroMQ>

=item L<Message::Passing::Output::WebHooks>

=item L<Message::Passing::Output::Search::Elasticsearch>

=item L<Message::Passing::Output::Redis>

=item L<Message::Passing::Output::Test>

=back

=head1 SEE ALSO

=over

=item L<Message::Passing::Manual> - The manual (contributions cherished).

=item L<http://www.slideshare.net/bobtfish/messaging-interoperability-and-log-aggregation-a-new-framework> - Slide deck!

=item L<Log::Message::Structured> - For creating your log messages.

=item L<Log::Dispatch::Message::Passing> - use Message::Passing outputs from L<Log::Dispatch>.

=back

=head1 THIS MODULE

This is a simple L<MooX::Options> script, with one input, one filter
and one output. To build your own similar scripts, see:

=over

=item L<Message::Passing::DSL> - To declare your message chains.

=item L<Message::Passing::Role::CLIComponent> - To provide C<foo> and C<foo_options> attribute pairs.

=item L<Message::Passing::Role::Script> - To provide daemonization features.

=back

=head2 METHODS

=head3 build_chain

Builds and returns the configured chain of input => filter => output.

=head3 start

Class method to call the run_message_server function with the results of
having constructed an instance of this class, parsed command line options
and constructed a chain.

This is the entry point for the script.

=head1 AUTHOR

Tomas (t0m) Doran <bobtfish@bobtfish.net>

=head1 SUPPORT

=head2 Bugs

Please log bugs at L<rt.cpan.org>. Each distribution has a bug tracker
link in its L<metacpan.org> page.

=head2 Discussion

L<#message-passing> on L<irc.perl.org>.

=head2 Source code

Source code for all modules is available at L<http://github.com/suretec>
and forks/patches are very welcome.

=head1 SPONSORSHIP

This module exists due to the wonderful people at Suretec Systems Ltd.
<http://www.suretecsystems.com/> who sponsored its development for its
VoIP division called SureVoIP <http://www.surevoip.co.uk/> for use with
the SureVoIP API -
<http://www.surevoip.co.uk/support/wiki/api_documentation>

=head1 COPYRIGHT

Copyright Suretec Systems Ltd. 2012.

Logstash (upon which many ideas for this project are based, but
from which we do not reuse any code) is copyright 2010 Jorden Sissel.

=head1 LICENSE

GNU Library General Public License, Version 2.1

=cut

